# Day 06: Chronal Coordinates
[link](https://adventofcode.com/2018/day/6)

## Part 1: Biggest finite area of the points closest to one of the given coordinates

In [1]:
import numpy as np

DTYPE = np.int16

# Load the grid which allows to place all the points.
# Then fill the grid with the areas of the points closest to some of the given coordinates
def load_grid(coords):
  mx, my = np.max(coords[:,0])+1, np.max(coords[:,1])+1
  grid = np.full(shape=(mx,my), fill_value=-1, dtype=DTYPE)
  for x in range(mx):
    for y in range(my):
      # Mahnattan distances from the (x,y) to all the points
      distances = [abs(x-px) + abs(y-py) for (px,py) in coords]
      closest_distance = min(distances)
      if distances.count(closest_distance) == 1:  # There is only one closest point
        closest_point_idx = distances.index(closest_distance)
        grid[x,y] = closest_point_idx  # Point ID on the grid is 
  return grid

def largest_area(grid):
  mx,my = grid.shape

  # IDs of all points on the grid.
  # Negative values are excluded because they represent points equally distant to two or more incoming coordinates
  point_ids = np.unique(grid[grid>=0])

  # IDs of all areas which touch grid edges.
  # Those areas will expand to infinity, so they will be later excluded from area calucuiation
  edge_ids = np.unique(np.concatenate([grid[0,:], grid[mx-1,:], grid[:,0], grid[:,my-1]]))

  # IDs of areas that are finite, i.e. don't expand to infinity
  finite_area_ids = [id for id in point_ids if id not in edge_ids]
#   print(''.join(char_map[i] for i in non_edge_area_ids))
  non_edge_areas = [np.count_nonzero(grid==id) for id in finite_area_ids]
#   print(non_edge_areas)
  return max(non_edge_areas)

In [2]:

test_in = np.array([(1, 1), (1, 6), (8, 3), (3, 4), (5, 5), (8, 9)], DTYPE)
test_grid = load_grid(test_in)

CHR_MAP = [chr(ord('a') + i) for i in range(26)]
for row in test_grid:
  print(''.join(CHR_MAP[id] if id>=0 else '.' for id in row))

assert largest_area(test_grid) == 17

aaaa.bbbbb
aaaa.bbbbb
aaadd.bbbb
aadddd....
aadddeeeef
..eeeeeeef
cccceeeeff
ccccceefff
cccccc.fff


In [3]:
import re

with open('06 input.txt', 'r') as file:
  puzzle_input = np.array([tuple(map(int, re.findall('-?\d+', line))) for line in file], DTYPE)

puzzle_input.shape

(50, 2)

In [4]:
largest_area(load_grid(puzzle_input))

5035

**Part 1 correct answer**: `5035`

## Part 2: Area with points whose distance is less than given

In [5]:
def close_area(coords, max_total_dist):
  mx, my = np.max(coords[:,0])+1, np.max(coords[:,1])+1
  grid = np.zeros(shape=(mx,my), dtype=DTYPE)
  for x in range(mx):
    for y in range(my):
      # Mahnattan distances from the (x,y) to all the points
      distances = [abs(x-px) + abs(y-py) for (px,py) in coords]
      if sum(distances) < max_total_dist:
        grid[x,y] = 1
  return np.count_nonzero(grid==1)

assert close_area(test_in, 32) == 16

close_area(puzzle_input, 10000)

35294

**Part 2 correct answer:** `35294`